retworkx-core/src/shortest_path/dijkstra.rs

```diff
@@ -111,63 +108,31 @@ where
     F: FnMut(G::EdgeRef) -> Result<K, E>,
     K: Measure + Copy,
 {
-    let mut visited = graph.visit_map();
     let mut scores = DictMap::new();
-    let mut visit_next = BinaryHeap::new();
-    let zero_score = K::default();
-    scores.insert(start, zero_score);
-    visit_next.push(MinScored(zero_score, start));
-    if path.is_some() {
-        path.as_mut().unwrap().insert(start, vec![start]);
+    if let Some(ref mut path) = path {
+        path.insert(start, vec![start]);
     }
-    while let Some(MinScored(node_score, node)) = visit_next.pop() {
-        if visited.is_visited(&node) {
-            continue;
-        }
-        if goal.as_ref() == Some(&node) {
-            break;
-        }
-        for edge in graph.edges(node) {
-            let next = edge.target();
-            if visited.is_visited(&next) {
-                continue;
-            }
-            let cost = edge_cost(edge)?;
-            let next_score = node_score + cost;
-            match scores.entry(next) {
-                Occupied(ent) => {
-                    if next_score < *ent.get() {
-                        *ent.into_mut() = next_score;
-                        visit_next.push(MinScored(next_score, next));
-                        if path.is_some() {
-                            let mut node_path = path
-                                .as_mut()
-                                .unwrap()
-                                .get(&node)
-                                .unwrap()
-                                .clone();
-                            node_path.push(next);
-                            path.as_mut().unwrap().entry(next).and_modify(
-                                |new_vec| {
-                                    *new_vec = node_path;
-                                },
-                            );
-                        }
-                    }
+
+    dijkstra_search(graph, Some(start), edge_cost, |event| {
+        match event {
+            DijkstraEvent::Discover(node, score) => {
+                scores.insert(node, score);
```

Comment:
I would rather keep the code duplicated and straightforward for our Dijksta implementation. I do not think we get an advantage with the dijkstra_search rewrite, in fact I found that the code has more "magic" hiding the implementation details. Also, performance could be affected

Reply:
Yeah, sure, I can revert the changes. I was just trying to avoid code duplication. However, if using `dijkstra_search` feels a bit unnatural and it's always more straightforward to rewrite Dijkstra's algorithm, I'm starting to doubt about the benefits of this series of PRs. My intention was to provide Python/Rust users or internal code the ability to customize the logic of common graph algorithms without the need to rewrite implementation details. Any performance benefits was a secondary concern. This means if you find this code less readable, which I totally get it, I have second thoughts about these PRs.

Reply:
The PRs are useful. For end-users in Rust the code is convenient, re-implementing Dijkstra is a pain because they have to worry about priority queues, comparing elements, correctness, and other things. For Python users it has all the conveniences from before, and using the code from the PR is much faster than writting it in Python.

The thing is `retworkx` is a special case because we already have our Dijkstra code written and working. I commented more to debate: should we really change what is already working?

If performance is not affected, keeping the shorter version is fine for me.

Reply:
@mtreinish Could you please run https://github.com/mtreinish/retworkx-comparison-benchmarks with this branch to see if there is a performance regression?

Reply:
I've run the benchmark suite for shortest path, there is a noticeable regression with this PR:

![single_source_shortest_path_2](https://user-images.githubusercontent.com/2447371/143048804-9e3f3e60-8fe3-41d6-b94a-026acdf7f655.png)
![single_source_shortest_path](https://user-images.githubusercontent.com/2447371/143048806-30e178ca-8c77-4a4c-8878-1ecfaf158ed3.png)
![all_pairs](https://user-images.githubusercontent.com/2447371/143048808-02af5baf-c974-4c04-a246-adbca7dba9a7.png)

What I'm not sure of though is how much `main` differs from `0.10.2` though, because as was discussed here the use of `DictMap` hasn't been released yet which is slightly slower than just hashbrown's hashmap.

Reply:
I ran it with main too so we could compare. There is a regression with main on the single source benchmarks too. I think we might want to revisit our `DictMap` usage everywhere and only use it places where we think iteration order actually matters (I don't think dijkstra's output is one of those, at least for retworkx-core). But the regression with this PR is worse: 

![single_source_shortest_path_2](https://user-images.githubusercontent.com/2447371/143054031-d6accd64-fd18-469a-af60-c3f38512490a.png)
![single_source_shortest_path](https://user-images.githubusercontent.com/2447371/143054032-c6d6e96e-c523-47e3-ad27-5f2e00fc6f91.png)
![all_pairs](https://user-images.githubusercontent.com/2447371/143054407-8e83dfa6-5543-444b-bf82-d957f22794ce.png)

Reply:
@mtreinish Thanks! Wow, didn't expect this huge performance regression! Reverted in 7e0e09f.